File: rust/src/day2.rs
use regex::Regex;
use std::cmp;
use std::collections::HashMap;

pub fn solution1(all_input: String) -> i32 {
    let mut result = 0;
    for line in all_input.lines() {
        result += get_valid_game_id(line.to_string());
    }
    result
}

fn get_valid_game_id(line: String) -> i32 {
    let limit_blocks = HashMap::from([("red", 12), ("green", 13), ("blue", 14)]);
    let re_game = Regex::new(r"Game\s+(\d+)").unwrap();
    let re_blocks = Regex::new(r"(\d+)\s+(red|green|blue)").unwrap();
    let game_id = re_game
        .captures(&line)
        .unwrap()
        .get(1)
        .unwrap()
        .as_str()
        .parse::<i32>()
        .unwrap();
    for blocks in line.split(";") {
        for color_block in re_blocks.captures_iter(&blocks) {
            let val_color = color_block.get(1).unwrap().as_str().parse::<i32>().unwrap();
            let name_color = color_block.get(2).unwrap().as_str();
            if val_color > limit_blocks[name_color] {
                return 0;
            }
        }
    }
    game_id
}

fn get_valid_game_id_with_min_blocks(line: String) -> i32 {
    let re_blocks = Regex::new(r"(\d+)\s+(red|green|blue)").unwrap();
    let mut min_blocks = HashMap::new();
    for blocks in line.split(";") {
        for color_block in re_blocks.captures_iter(&blocks) {
            let val_color = color_block.get(1).unwrap().as_str().parse::<i32>().unwrap();
            let name_color = color_block.get(2).unwrap().as_str();
            if let Some(value) = min_blocks.get_mut(name_color) {
                *value = cmp::max(*value, val_color);
            } else {
                min_blocks.insert(name_color, val_color);
            }
        }
    }
    min_blocks.values().product()
}

pub fn solution2(all_input: String) -> i32 {
    let mut result = 0;
    for line in all_input.lines() {
        result += get_valid_game_id_with_min_blocks(line.to_string());
    }
    result
}

File: cpp/day2.cpp
#include "day2.h"
#include <string>
#include <sstream>
#include <iterator>
#include <regex>
#include <unordered_map>

namespace day2{
int solution1(const std::string& allInput){
    std::istringstream iss(allInput);
    std::string line;
    int result = 0;
    while(std::getline(iss, line)){
        result += getValidGame(line);
    }
    return result;
}

int getValidGame(const std::string& line) {
    std::regex re_game("Game\\s+(\\d+)");
    std::smatch matchGame;
    std::regex_search(line, matchGame, re_game);
    int gameId = std::stoi(matchGame.str(1));

    std::regex re_blocks("(\\d+)\\s+(red|green|blue)");
    std::stringstream lineStrem(line);
    std::string block;
    std::vector<std::string> listBlocks;

    std::pmr::unordered_map<std::string, int> colorBlocks = {
        {"red", 12},
        {"green", 13},
        {"blue", 14},
    };

    while(std::getline(lineStrem, block, ';')){
        std::smatch matchBlock;
        for (std::sregex_iterator it = std::sregex_iterator(block.begin(), block.end(), re_blocks);
     it != std::sregex_iterator(); ++it){
            std::smatch colVal = *it;
            int colorValue = std::stoi(colVal.str(1));
            if(colorValue > colorBlocks[colVal.str(2)]){
                return 0;
            }
        }
    }
    return gameId;
}

int solution2(const std::string& allInput) {
    std::istringstream iss(allInput);
    std::string line;
    int result = 0;
    while(std::getline(iss, line)){
        result += getPowerBlocks(line);
    }
    return result;
}

int getPowerBlocks(const std::string& line) {
    std::pmr::unordered_map<std::string, int> colorBlocks = {
        {"red", 0},
        {"green", 0},
        {"blue", 0},
    };
    std::regex re_blocks("(\\d+)\\s+(red|green|blue)");
    std::stringstream lineStream(line);
    std::string block;
    std::vector<std::string> listBlocks;
    while(std::getline(lineStream, block, ';')){
        std::smatch matchBlock;
        for (std::sregex_iterator it = std::sregex_iterator(block.begin(), block.end(), re_blocks);
     it != std::sregex_iterator(); ++it){
            std::smatch colVal = *it;
            colorBlocks[colVal.str(2)] = std::max(std::stoi(colVal.str(1)), colorBlocks[colVal.str(2)]);
        }
    }
    return colorBlocks["red"] * colorBlocks["green"] * colorBlocks["blue"];
}

}
